src/client/texture_atlas.cpp

```diff
@@ -0,0 +1,236 @@
+/*
+Minetest
+Copyright (C) 2024 Andrey, Andrey2470T <andreyt2203@gmail.com>
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "client/texture_atlas.h"
+#include "settings.h"
+#include "log.h"
+
+
+TextureAtlas::TextureAtlas(video::IVideoDriver *vdrv, ITextureSource *src, std::vector<TileLayer*> &layers)
+{
+	m_driver = vdrv;
+
+	m_tsrc = src;
+
+	u32 atlas_size = 0;
+
+	// Tile layers with unique ITextures
+	std::vector<TileLayer *> unique_layers;
+
+	for (auto layer : layers) {
+		// If for the layer the TileInfo was already created,
+		// just save the index at this TileInfo and omit the iteration
+		bool is_layer_collected = false;
+		for (u32 i = 0; i < unique_layers.size(); i++)
+			if (unique_layers[i]->texture == layer->texture) {
+				layer->atlas_tile_info_index = i;
+				is_layer_collected = true;
+				break;
+			}
+
+		if (is_layer_collected)
+			continue;
+
+		TileInfo tile_info;
+
+		if (layer->animation_frame_count > 1)
+			tile_info.tex = (*layer->frames)[0].texture;
+		else
+			tile_info.tex = layer->texture;
+
+		if (tile_info.tex) {
+			core::dimension2du size = tile_info.tex->getSize();
+			tile_info.width = size.Width;
+			tile_info.height = size.Height;
+		}
+
+		if (layer->animation_frame_count > 1) {
+			tile_info.anim.frame_length_ms = layer->animation_frame_length_ms;
+			tile_info.anim.frame_count = layer->animation_frame_count;
+
+			//if (g_settings->getBool("desynchronize_mapblock_texture_animation"))
+			//	tile_info.anim.frame_offset =
+			//			100000 * (2.0 + noise3d(
+			//			blockpos.X, blockpos.Y,
+			//			blockpos.Z, 0));
+
+			for (auto &frame : *layer->frames)
+				tile_info.anim.frames.push_back(frame.texture);
+		}
+
+		layer->atlas_tile_info_index = m_tiles_infos.size();
+		m_tiles_infos.push_back(tile_info);
+		unique_layers.push_back(layer);
+
+		atlas_size += (tile_info.width * tile_info.height);
+	}
+
+	u32 atlas_side = std::min(getClosestPowerOfTwo(std::sqrt((f32)atlas_size)), 16384u);
+
+	packTextures(atlas_side);
+
+	// Extending the atlas width twice for allocating the additional texture space for cracks
+	m_atlas_texture = m_driver->addTexture(core::dimension2du(atlas_side*2, atlas_side), "DiffuseAtlas", video::ECF_A32B32G32R32F);
```

Comment:
this dimensions is a worst case bound right?  Could check the actual size of the packed texture and use that.

Reply:
Fixed also.

Reply:
I was asking if `core::dimension2du(atlas_side*2, atlas_side)` is a worst case for the atlas dimensions, and suggesting that the actual max width and height is used (can be calculated in `packTextures`). 

I see the change made the `GL_MAX_TEXTURE_SIZE` worst case bound was replaced with the actual value, was I just mistaken about the atlas dimensions side of things? Will it always need exactly that size texture?


On a general note, I think it would be cleaner in any case if `packTextures` returned the bounds it needs. Just cause the bounds depend on *how* the textures are packed.